Add support for anyOf and type: null

allenporter opened the pull request:
Add support for the `anyOf` openapi type. This also changes the behavior of nullable fields, prefering anyof  with a none type. The validation tests that exercise the combinations of schemas and converters has been updated to verify compatibility between openapi and voluptuous for these more specific scenarios.

Fixes here are meant to address report in https://github.com/home-assistant/core/issues/144300

Shulyaka commented:
I didn't know this type existed. Is it s new addition to the standard?

Shulyaka commented:
It's a shame they do backwards-incompatible changes with a minor version bump!
May be we should add an additional parameter to set the target OpenAPI version? Or just test all the supported consumers if they already support OpenAPI 3.1? (which I doubt as this could require an expensive fine-tuning of some models)

allenporter commented:
I've updated this to output the old and new version. I've updated the compatibility tests so they always use the new version, so its nice to just have the 4 way comparisons of formats without exceptions.

Really the motivation is just to accept 3.1 and convert it to voluptuous so this should be fine.

Shulyaka reviewed (approved):
Looks good to me
